quarkchain/evm/messages.py

```diff
@@ -129,57 +131,60 @@ def validate_transaction(state, tx):
     if tx.startgas < total_gas:
         raise InsufficientStartGas(rp(tx, "startgas", tx.startgas, total_gas))
 
-    # (4.0) require transfer_token_id and gas_token_id to be in allowed list
-    if tx.transfer_token_id not in state.qkc_config.allowed_transfer_token_ids:
-        raise InsufficientBalance(
-            "{}: token {} is not in allowed transfer_token list".format(
-                tx.__repr__(), token_id_decode(tx.transfer_token_id)
+    default_chain_token = state.shard_config.default_chain_token
+    bal = {tx.transfer_token_id: state.get_balance(tx.sender, tx.transfer_token_id)}
+    if tx.transfer_token_id != tx.gas_token_id:
+        bal[tx.gas_token_id] = state.get_balance(tx.sender, tx.gas_token_id)
+
+    # (4) requires non-zero balance for transfer_token_id and gas_token_id if non-default
+    for token_id in [tx.transfer_token_id, tx.gas_token_id]:
```

Comment:
what is the purpose of this?  Looks like this should be a protocol of the upper layer.

Reply:
as discussed offline, this will prevent senders using arbitrary token IDs to confuse other users (e.g. a tx with non-existent token id KLQJDLASD)